File: src/marsskfas.cpp
#include "marsskfas.h"
#include "genutils.h"
#include "marssutils.h"

using namespace arma;
using namespace Rcpp;

List marsskfas_impl(List& MLEobj, bool only_logLik, bool return_lag_one,
		    bool return_kfas_model)
{
  List modelObj = MLEobj["marss"];
  List control = MLEobj["control"];
  bool diffuse = modelObj["diffuse"];
  List model_dims = modelObj.attr("model.dims");

  int n = as<IntegerVector>(model_dims["data"])[0];
  int TT = as<IntegerVector>(model_dims["data"])[1];
  int m = as<IntegerVector>(model_dims["x"])[0];

  int g1 = as<IntegerVector>(model_dims["Q"])[0];
  int h1 = as<IntegerVector>(model_dims["R"])[0];
  int l1 = as<IntegerVector>(model_dims["L"])[0];

  List par_1 = parmat(MLEobj);

  std::map<std::string, arma::mat> matpar_1;
  if (!RObject(par_1.attr("names")).isNULL()) {
    CharacterVector parnames = par_1.attr("names");
    for (String parname : parnames) {
      matpar_1[parname] = as<mat>(as<NumericMatrix>(par_1[parname]));
    }
  }

  mat t_B = matpar_1["B"].t();
  
  IntegerVector TTindices(TT);
  IntegerVector TTindices1(TT-1);
  for (int i = 0; i < TT; ++i) {
    TTindices[i] = i;
    if (i < TT - 1) {
      TTindices1[i] = i + 1;
    }
  }

  cube Zt;
  cube stack_Zt;
  
  if (as<IntegerVector>(model_dims["Z"])[2] == 1 &&
      as<IntegerVector>(model_dims["A"])[2] == 1) {
    mat& mAt = matpar_1["A"];
    mat& mZt = matpar_1["Z"];
    Zt = cube(mAt.n_rows, mAt.n_cols + mZt.n_cols, 1);
    Zt.subcube(0, 0, 0, mZt.n_rows-1, mZt.n_cols-1, 0) = mZt;
    Zt.subcube(0, mZt.n_cols, 0, mZt.n_rows-1, mAt.n_cols + mZt.n_cols - 1, 0) =
      mAt;
    stack_Zt = cube(n, 2*(m+1), 1, fill::zeros);
    stack_Zt.subcube(0, 0, 0, n-1, m, 0) = Zt;
  }
  else {
    Zt = cube(n, m+1, TT, fill::zeros);
    stack_Zt = cube(n, 2*(m+1), TT, fill::zeros);
    std::map<std::string, cube> pars = parmat_cube(MLEobj, {"Z", "A"}, TTindices);
    Zt.subcube(0, 0, 0, n-1, m-1, TT-1) = pars["Z"];
    Zt.subcube(0, m, 0, n-1, m, TT-1) = pars["A"];
    stack_Zt.subcube(0, 0, 0, n-1, m, TT-1) = Zt;
  }

  cube Tt;
  cube stack_Tt;

  if (as<IntegerVector>(model_dims["B"])[2] == 1 &&
      as<IntegerVector>(model_dims["U"])[2] == 1) {

    Tt = cube(m + 1, m + 1, 1, fill::zeros);
    Tt.subcube(0, 0, 0, m-1, m-1, 0) = matpar_1["B"];
    Tt.subcube(0, m, 0, m-1, m, 0) = matpar_1["U"];
    Tt(m, m, 0) = 1;

    stack_Tt = cube(2*(m+1), 2*(m+1), 1, fill::zeros);
    stack_Tt.subcube(0, 0, 0, m, m, 0) = Tt;
    mat dummy_eye(m+1, m+1, fill::eye);
    stack_Tt.subcube(m+1, 0, 0, 2*m+1, m, 0) = dummy_eye;
  }
  else {
    Tt = cube(m+1, m+1, TT, fill::zeros);
    stack_Tt = cube(2*(m+1), 2*(m+1), TT, fill::zeros);

    std::map<std::string,cube>  pars = parmat_cube(MLEobj, {"B", "U"}, TTindices);
    Tt.subcube(0, 0, 0, m-1, m-1, TT-2) =
      pars["B"].subcube(0, 0, 1, m-1, m-1, TT-1);
    Tt.subcube(0, m, 0, m-1, m, TT-2) =
      pars["U"].subcube(0, 0, 1, m-1 , 0, TT-1);
    Tt.subcube(m, m, 0, m, m, TT-1) = cube(1, 1, TT, fill::ones);
    Tt.subcube(0, 0, TT-1, m, m, TT-1) = mat(m+1, m+1, fill::ones);

    for (int i = 0; i < TT; ++i) {
      stack_Tt.subcube(m+1, 0, i, 2*m+1, m, i) = eye<mat>(m+1, m+1);
    }
    stack_Tt.subcube(0, 0, 0, m, m, TT-1) = Tt;
  }

  cube Ht;

  if (as<IntegerVector>(model_dims["R"])[2] == 1) {
    mat& mHt = matpar_1["H"];
    mat& mRt = matpar_1["R"];
    mat mtcp = (mHt * mRt) * mHt.t();
    Ht = cube(mtcp.n_rows, mtcp.n_cols, 1);
    Ht.slice(0) = mtcp;
  }
  else {
    Ht = cube(n, n, TT);
    cube tmpHcube = parmat_cube(MLEobj, {"H"}, TTindices)["H"];
    cube tmpRcube = parmat_cube(MLEobj, {"R"}, TTindices)["R"];
    for (int i = 0; i < TT; ++i) {
      mat& mHt = tmpHcube.slice(i);
      mat& mRt = tmpRcube.slice(i);
      Ht.slice(i) = (mHt * mRt) * mHt.t();
    }
  }

  cube Qt;
  cube stack_Qt;

  if (as<IntegerVector>(model_dims["Q"])[2] == 1) {
    Qt = cube(g1, g1, 1);
    Qt.slice(0) = matpar_1["Q"];
    stack_Qt = cube(2*g1, 2*g1, 1, fill::zeros);
    stack_Qt.subcube(0, 0, 0, g1-1, g1-1, 0) = Qt;
  }
  else {
    Qt = cube(g1, g1, TT, fill::zeros);
    stack_Qt = cube(2*g1, 2*g1, TT, fill::zeros);

    Qt.subcube(0, 0, 0, g1-1, g1-1, TT-2) =
      parmat_cube(MLEobj, "Q", TTindices1)["Q"];
    stack_Qt.subcube(0, 0, 0, g1-1, g1-1, TT-2) =
      Qt.subcube(0, 0, 0, g1-1, g1-1, TT-2) ;
  }

  cube Rt;
  cube stack_Rt;
  if (as<IntegerVector>(model_dims["R"])[2] == 1) {
    Rt = cube(m+1, g1, 1, fill::zeros);
    stack_Rt = cube(2*(m+1), 2*g1, 1, fill::zeros);
    
    Rt.subcube(0, 0, 0, m-1, g1-1, 0) = matpar_1["G"];
    stack_Rt.subcube(0, 0, 0, m, g1-1, 0) = Rt;
  }
  else {
    Rt = cube(m+1, g1, TT, fill::zeros);
    stack_Rt = cube(2*(m+1), 2*g1, TT, fill::zeros);

    cube tmppar=parmat_cube(MLEobj, "G", TTindices1)["G"];

    Rt.subcube(0, 0, 0, m-1, g1-1, TT-2) =
      parmat_cube(MLEobj, "G", TTindices1)["G"];
    stack_Rt.subcube(0, 0, 0, m, g1-1, TT-1) = Rt;
  }
  
  mat x00;
  mat V00;
  mat x10;
  mat V10;

  double testtinitx = modelObj["tinitx"];

  if (testtinitx == 0) {
    x00 = matpar_1["x0"];
    V00 = (matpar_1["L"] * matpar_1["V0"]) * matpar_1["L"].t();
    x10 = matpar_1["B"] * x00 + matpar_1["U"];
    V10 = matpar_1["B"] * V00 * t_B + matpar_1["Q"];
  }
  else {
    x10 = matpar_1["x0"];
    x00 = mat(10, 1, fill::zeros);
    V10 = matpar_1["V0"];
    V00 = zeros(m, m);
  }

  mat onesrow(1, x10.n_cols, fill::ones);
  mat a1 = join_cols(x10, onesrow);
  mat stack_a1 = join_cols(join_cols(join_cols(x10, onesrow),x00),onesrow);

  mat P1inf(m+1, m+1, fill::zeros);
  mat stack_P1inf(2*(m+1), 2*(m+1), fill::zeros);
  mat P1(m+1, m+1, fill::zeros);
  mat stack_P1(2*(m+1), 2*(m+1), fill::zeros);
  
  if (diffuse) {
    P1inf.submat(0, 0, m-1, m-1) = V10;
    stack_P1inf.submat(0, 0, m-1, m-1) = V10;
    stack_P1inf.submat(m+1, m+1, 2*m, 2*m) = V00;
  }
  else {
    P1.submat(0, 0, m-1, m-1) = V10;
    stack_P1.submat(0, 0, m-1, m-1) = V10;
    stack_P1.submat(m+1, m+1, 2*m, 2*m) = V00;
    stack_P1.submat(0, m+1, m-1, 2*m) = matpar_1["B"] * V00;
    stack_P1.submat(m+1, 0, 2*m, m-1) = V00 * t_B;
  }

  List kfas_model;
  List envlist = List::create(
    _["a1"]=a1,
    _["control"]=control,
    _["diffuse"]=diffuse,
    _["Ht"]=Ht,
    // _["i"]=i,
    _["m"]=m,
    _["MLEobj"]=MLEobj,
    _["model.dims"]=model_dims,
    _["modelObj"]=modelObj,
    _["n"]=n,
    _["only.logLik"]=only_logLik,
    _["P1"]=P1,
    _["P1inf"]=P1inf,
    _["par.1"]=par_1,
    _["Qt"]=Qt,
    _["return.kfas.model"]=return_kfas_model,
    _["return.lag.one"]=return_lag_one,
    _["Rt"]=Rt
    );

  envlist.push_back(stack_a1,"stack.a1");
  envlist.push_back(stack_P1,"stack.P1");
  envlist.push_back(stack_P1inf,"stack.P1inf");
  envlist.push_back(stack_Qt,"stack.Qt");
  envlist.push_back(stack_Rt,"stack.Rt");
  envlist.push_back(stack_Tt,"stack.Tt");
  envlist.push_back(stack_Zt,"stack.Zt");
  envlist.push_back(t_B,"t.B");
  envlist.push_back(Tt,"Tt");
  envlist.push_back(TT,"TT");
  envlist.push_back(V00,"V00");
  envlist.push_back(V10,"V10");
  envlist.push_back(x00,"x00");
  envlist.push_back(x10,"x10");

  NumericMatrix nmyt = clone(as<NumericMatrix>(modelObj["data"]));
  NumericVector nvyt(nmyt);
  nvyt = ifelse(is_na(nvyt), NA_REAL, nvyt);
  nmyt = transpose(nmyt);
  envlist.push_back(nmyt, "yt");
  envlist.push_back(Zt,"Zt");

  kfas_model = SSModel3(envlist);

  if (only_logLik) {
    return List::create(_["logLik"]=logLik(kfas_model));
  }

  vec diag_R = diagvec(matpar_1["R"]);

  List ks_out = KFS(kfas_model, false);

  if (packageVersion("KFAS") != "0.9.11") {
    ks_out["a"] = transpose(as<NumericMatrix>(ks_out["a"]));
    ks_out["alphahat"] = transpose(as<NumericMatrix>(ks_out["alphahat"]));
  }
  NumericMatrix ks_out_a = ks_out["a"];
  NumericMatrix ks_out_alphahat = clone(as<NumericMatrix>(ks_out["alphahat"]));

  cube ksoV(as<cube>(ks_out["V"]));
  cube cVtT = ksoV.tube(0, 0, m-1, m-1);
  NumericVector VtT = wrap(cVtT);
  
  cube ksoP(as<cube>(ks_out["P"]));
  cube cVtt1 = ksoP.subcube(0, 0, 0, m-1, m-1, TT-1);
  NumericVector Vtt1 = wrap(cVtt1);

  NumericVector Vtt1T;
  if (return_lag_one) {
    cube cVtt1T = ksoV.tube(0, m+1, m-1, 2*m);
    Vtt1T = wrap(cVtt1T);
  }

  if (any(diag_R == 0)) {
    VtT = ifelse(sapply(VtT,[](double x){return x < datum::eps;}), 0, VtT);
    Vtt1 = ifelse(sapply(Vtt1,[](double x){return x < datum::eps;}), 0, Vtt1);
    if (return_lag_one) {
      Vtt1T = ifelse(sapply(Vtt1T,[](double x){return x < datum::eps;}), 0, Vtt1T);
    }
  }

  NumericVector x10T = ks_out_alphahat.column(0);
  x10T.erase(m, ks_out_alphahat.nrow());
  IntegerVector x10Tdims {m, 1};
  x10T.attr("dim") = x10Tdims;
  
  NumericVector x0T;
  
  NumericMatrix V10T = wrap(cVtT.slice(0));
  NumericMatrix V0T;
  
  if (testtinitx == 1) {
    x00 = mat(m, 1);
    x00.fill(NA_REAL);
    V00 = mat(m, m);
    V00.fill(NA_REAL);
    x0T = x10T;
    V0T = V10T;
  }
  else {
    NumericMatrix Vtt1_1 = sub3D(Vtt1,0);
    mat mVtt1_1(Vtt1_1.begin(), Vtt1_1.nrow(), Vtt1_1.ncol(), false);
    mat Vinv = pcholinv(mVtt1_1);
    if (m != 1) {
      Vinv = (Vinv.t() + Vinv) / 2.;
    }
    mat J0 = V00 * t_B * Vinv;
    NumericVector ks_out_a_col = ks_out_a.column(0);
    int erase_end = ks_out_a_col.length();
    ks_out_a_col.erase(m, erase_end);

    vec t1 = as<vec>(x10T) - as<vec>(ks_out_a_col);
    mat t2 = J0 * t1;
    x0T = x00 + t2;
    V0T = wrap(V00 + J0 * (cVtT.slice(0) - cVtt1.slice(0)) * J0.t());
  }

  List rtn_list = List::create(
    _["xtT"] = add_default_rownames(
      ks_out_alphahat(Range(0, m-1), Range(0, ks_out_alphahat.ncol()-1))),
    _["VtT"] = VtT,
    _["Vtt1T"] = return_lag_one ? wrap(Vtt1T) : R_NilValue,
    _["x0T"] = x0T,
    _["V0T"] = V0T,
    _["x10T"] = add_default_rownames(NumericMatrix(x10T.length(), 1, x10T.begin())),
    _["V10T"] = V10T,
    _["x00T"] = x00,
    _["V00T"] = V00,
    _["Vtt"] = String("Use MARSSkfss to get Vtt"),
    _["Vtt1"] = Vtt1,
    _["J"] = String("Use MARSSkfss to get J"),
    _["J0"] = String("Use MARSSkfss to get J0"),
    _["Kt"] = String("Use MARSSkfss to get Kt"),
    _["xtt1"] = add_default_rownames(ks_out_a(Range(0, m-1), Range(0,TT-1))),
    _["xtt"] = String("Use MARSSkfss to get xtt"),
    _["Innov"] = String("Use MARSSkfss to get Innov"),
    _["Sigma"] = String("Use MARSSkfss to get Sigma"),
    _["kfas.model"] = return_kfas_model ? wrap(kfas_model) : R_NilValue,
    _["logLik"] = ks_out["logLik"]
    );

  rtn_list.push_back(true, "ok");
  rtn_list.push_back(R_NilValue, "errors");

  return rtn_list;
}

File: R/MARSSkfas.r
#######################################################################################################
#   Kalman filter and smoother function based on Koopman and Durbin's KFAS package
#   y_t = Z_t * alpha_t + eps_t (observation equation)
#   alpha_t+1 = T_t * alpha_t + R_t * eta_t(transition equation)
#   Note the different time indexing for transition equation
#   The y_t,y_{t-1} stacked vector with parameters vectors reconfigured to output the
#     smoothed lag-1 covariances needed by the EM algorithm
#######################################################################################################
MARSSkfas <- function(MLEobj, only.logLik = FALSE, return.lag.one = TRUE, return.kfas.model = FALSE) {
  MODELobj <- MLEobj[["marss"]]
  control <- MLEobj[["control"]]
  diffuse <- MODELobj[["diffuse"]]
  model.dims <- attr(MODELobj, "model.dims")
  
  n <- model.dims$data[1]
  TT <- model.dims$data[2]
  m <- model.dims$x[1]
  g1 <- model.dims$Q[1]
  h1 <- model.dims$R[1]
  l1 <- model.dims$L[1]
  par.1 <- parmat(MLEobj, t = 1)
  t.B <- matrix(par.1$B, m, m, byrow = TRUE)
  # create the YM matrix
  YM <- matrix(as.numeric(!is.na(MODELobj$data)), n, TT)
  
  # Make sure the missing vals in yt are NAs if there are any
  yt <- MODELobj$data
  yt[!YM] <- as.numeric(NA)
  
  # Stack the y so that we can get the lag-1 covariance smoother from the Kalman filter output
  # stack.yt=rbind(yt[,1:TT,drop=FALSE],cbind(NA,yt[,1:(TT-1),drop=FALSE]))
  # stack.yt=t(stack.yt)
  
  # KFS needs time going down rows so yt can be properly converted to ts object
  yt <- t(yt)
  
  # Build the Zt matrix which is n x (m+1) or n x (m+1) x T; (m+1) because A is in Z
  if (model.dims$Z[3] == 1 & model.dims$A[3] == 1) {
    # not time-varying
    Zt <- cbind(par.1$Z, par.1$A)
    stack.Zt <- matrix(0, n, 2 * (m + 1))
    stack.Zt[1:n, 1:(m + 1)] <- Zt
  } else {
    Zt <- array(0, dim = c(n, m + 1, TT))
    stack.Zt <- array(0, dim = c(n, 2 * (m + 1), TT))
    pars <- parmat(MLEobj, c("Z", "A"), t = 1:TT)
    Zt[1:n, 1:m, ] <- pars$Z
    Zt[1:n, m + 1, ] <- pars$A
    stack.Zt[1:n, 1:(m + 1), ] <- Zt
  }
  
  # Build the Tt matrix which is (m+1) x (m+1) or (m+1) x (m+1) x T; (m+1) because x has extra row of 1s (for the A)
  #    Tt=cbind(rbind(parList$B,matrix(0,1,m)),matrix(c(parList$U,1),m+1,1))
  if (model.dims$B[3] == 1 & model.dims$U[3] == 1) {
    # not time-varying
    Tt <- cbind(rbind(par.1$B, matrix(0, 1, m)), matrix(c(par.1$U, 1), m + 1, 1))
    stack.Tt <- matrix(0, 2 * (m + 1), 2 * (m + 1))
    stack.Tt[1:(m + 1), 1:(m + 1)] <- Tt
    stack.Tt[(m + 2):(2 * m + 2), 1:(m + 1)] <- diag(1, m + 1)
  } else {
    Tt <- array(0, dim = c(m + 1, m + 1, TT))
    stack.Tt <- array(0, dim = c(2 * (m + 1), 2 * (m + 1), TT))
    pars <- parmat(MLEobj, c("B", "U"), t = 1:TT)
    # pars uses i+1 because in KFAS x(t)=B(t-1)x(t-1) versus MARSS where x(t)=B(t)x(t-1)
    # KFAS sets prior on x(1) so recursions start at x(2); thus B(0) never appears in the KFAS recursions
    Tt[1:m, 1:m, 1:(TT - 1)] <- pars$B[, , 2:TT]
    Tt[1:m, m + 1, 1:(TT - 1)] <- pars$U[, , 2:TT]
    Tt[m + 1, m + 1, ] <- 1
    # in KFAS T[,,TT] is not used since x(T)=B(t-1)x(t-1) is the last computation
    # Set to 0 since is.SSModel does like NA in any parameters
    Tt[, , TT] <- 0
    stack.Tt[(m + 2):(2 * m + 2), 1:(m + 1), ] <- diag(1, m + 1)
    stack.Tt[1:(m + 1), 1:(m + 1), ] <- Tt
    stack.Tt[, , TT] <- 0 # see comment above re setting TT value
  }
  
  # Build the Ht (R) matrix which is n x n or n x n x T
  if (model.dims$R[3] == 1) {
    # not time-varying
    Ht <- tcrossprod(par.1$H %*% par.1$R, par.1$H)
  } else {
    Ht <- array(0, dim = c(n, n, TT))
    for (i in 1:TT) {
      Ht[, , i] <- tcrossprod(parmat(MLEobj, "H", t = i)$H %*% parmat(MLEobj, "R", t = i)$R, parmat(MLEobj, "H", t = i)$H)
    }
  }
  
  # Build the Qt matrix which is g1 x g1 or g1 x g1 x T;
  if (model.dims$Q[3] == 1) {
    # not time-varying
    Qt <- par.1$Q
    stack.Qt <- matrix(0, 2 * g1, 2 * g1)
    stack.Qt[1:g1, 1:g1] <- Qt
  } else {
    # See notes for Tt re differences in parameter indexing for the process equation
    Qt <- array(0, dim = c(g1, g1, TT))
    stack.Qt <- array(0, dim = c(2 * g1, 2 * g1, TT))
    for (i in 1:(TT - 1)) {
      # i+1 since in KFAS my B(t) equals B(t-1)
      Qt[, , i] <- parmat(MLEobj, "Q", t = i + 1)$Q
      stack.Qt[1:g1, 1:g1, i] <- Qt[, , i]
    }
    # see comments on setting of T re TT value; TT value never appears in the KFAS recursions
    # but is.SSModel check does not like any NAs in the parameters
    Qt[, , TT] <- 0
    stack.Qt[, , TT] <- 0
  }
  
  # build the Rt (G) matrix; process error is G(t)%*%w(t); extra row at bottom for for U (1s row in x)
  if (model.dims$G[3] == 1) {
    # not time-varying
    Rt <- matrix(0, m + 1, g1)
    Rt[1:m, 1:g1] <- par.1$G
    stack.Rt <- matrix(0, 2 * (m + 1), 2 * g1)
    stack.Rt[1:(m + 1), 1:g1] <- Rt
  } else {
    # See notes for Tt re differences in parameter indexing for the process equation
    Rt <- array(0, dim = c(m + 1, g1, TT))
    stack.Rt <- array(0, dim = c(2 * (m + 1), 2 * g1, TT))
    for (i in 1:(TT - 1)) {
      # i+1 since in KFAS my B(t) equals B(t-1)
      Rt[, , i] <- matrix(0, m + 1, g1)
      Rt[1:m, 1:g1, i] <- parmat(MLEobj, "G", t = i + 1)$G
      stack.Rt[1:(m + 1), 1:g1, i] <- Rt[, , i]
    }
    # see comments on setting of T re TT value; TT value never appears in the KFAS recursions
    # but is.SSModel check does not like any NAs in the parameters
    Rt[, , TT] <- 0
    stack.Rt[, , TT] <- 0
  }
  
  # Build the a1 and P1 matrices
  # First compute x10 and V10 if tinitx=0
  if (MODELobj$tinitx == 0) { # Compute needed x_1 | x_0
    # B(1),U(1), and Q(1) correct here since equivalent to B(0), etc in KFAS terminology
    x00 <- par.1$x0
    V00 <- tcrossprod(par.1$L %*% par.1$V0, par.1$L)
    x10 <- par.1$B %*% x00 + par.1$U # Shumway and Stoffer treatment of initial states
    V10 <- par.1$B %*% V00 %*% t.B + par.1$Q # eqn 6.20
  } else {
    x10 <- par.1$x0
    x00 <- matrix(10, m, 1) # dummy; not defined
    V10 <- par.1$V0
    V00 <- diag(0, m) # dummy; not defined
  }
  a1 <- rbind(x10, 1)
  stack.a1 <- rbind(x10, 1, x00, 1)
  if (diffuse) {
    P1inf <- matrix(0, m + 1, m + 1)
    P1inf[1:m, 1:m] <- V10
    stack.P1inf <- matrix(0, 2 * (m + 1), 2 * (m + 1))
    stack.P1inf[1:m, 1:m] <- V10
    stack.P1inf[(m + 2):(2 * m + 1), (m + 2):(2 * m + 1)] <- V00
    P1 <- matrix(0, m + 1, m + 1)
    stack.P1 <- matrix(0, 2 * (m + 1), 2 * (m + 1))
  } else {
    P1inf <- matrix(0, m + 1, m + 1)
    stack.P1inf <- matrix(0, 2 * (m + 1), 2 * (m + 1))
    P1 <- matrix(0, m + 1, m + 1)
    P1[1:m, 1:m] <- V10
    # P1 is the var-cov matrix for the stacked x1,x0
    # it is matrix(c(V10,B*V00,V00*t(B),V00),2,2)
    # x1=B*x0+U+w; in the var-cov mat looks like
    # E[x1*t(x1)] E[(Bx0+U+w1)*t(x0)] -   E[x1]*E[x1]     E[(Bx0+U+w1)]E[t(x0)]
    # E[x0*t(Bx0+U+w1)] E[x0*t(x0)]       E[(Bx0+U+w1)]E[t(x0)] - E[x0]E[t(x0)]
    stack.P1 <- matrix(0, 2 * (m + 1), 2 * (m + 1))
    stack.P1[1:m, 1:m] <- V10
    stack.P1[(m + 2):(2 * m + 1), (m + 2):(2 * m + 1)] <- V00
    stack.P1[1:m, (m + 2):(2 * m + 1)] <- par.1$B %*% V00
    stack.P1[(m + 2):(2 * m + 1), 1:m] <- tcrossprod(V00, par.1$B)
  }
  
  if (!return.lag.one) {
    kfas.model <- SSModel(yt ~ -1 + SSMcustom(Z = Zt, T = Tt, R = Rt, Q = Qt, a1 = a1, P1 = P1, P1inf = P1inf), H = Ht)
  } else {
    kfas.model <- SSModel(yt ~ -1 + SSMcustom(Z = stack.Zt, T = stack.Tt, R = stack.Rt, Q = stack.Qt, a1 = stack.a1, P1 = stack.P1, P1inf = stack.P1inf), H = Ht)
  }
  
  if (only.logLik) {
    return(list(logLik = logLik(kfas.model)))
  }
  # This works because I do not allow the location of 0s on the diagonal of R or Q to be time-varying
  if (n == 1) {
    diag.R <- unname(par.1$R)
  } else {
    diag.R <- unname(par.1$R)[1 + 0:(n - 1) * (n + 1)]
  }
  
  if (any(diag.R == 0)) { # because KFAS 1.0.4 added a warning message
    ks.out <- suppressWarnings(KFS(kfas.model, simplify = FALSE))
  } else {
    ks.out <- KFS(kfas.model, simplify = FALSE)
  } # simplify=FALSE so 1.0.0 returns LL
  # because 1.0.0 has time down columns instead of across rows
  if (packageVersion("KFAS") != "0.9.11") {
    ks.out$a <- t(ks.out$a)
    ks.out$alphahat <- t(ks.out$alphahat)
  }
  
  VtT <- ks.out$V[1:m, 1:m, , drop = FALSE]
  Vtt1 <- ks.out$P[1:m, 1:m, 1:TT, drop = FALSE]
  if (!return.lag.one) {
    Vtt1T <- NULL
  } else {
    Vtt1T <- ks.out$V[1:m, (m + 2):(2 * m + 1), , drop = FALSE]
  }
  # zero out rows cols as needed when R diag = 0
  # Check that if any R are 0 then model is solveable
  if (any(diag.R == 0)) {
    VtT[abs(VtT) < .Machine$double.eps] <- 0
    Vtt1[abs(Vtt1) < .Machine$double.eps] <- 0
    if (return.lag.one) Vtt1T[abs(Vtt1T) < .Machine$double.eps] <- 0
  }
  
  x10T <- ks.out$alphahat[1:m, 1, drop = FALSE]
  V10T <- matrix(VtT[, , 1], m, m)
  if (MODELobj$tinitx == 1) {
    x00 <- matrix(NA, m, 1)
    V00 <- matrix(NA, m, m)
    x0T <- x10T
    V0T <- V10T
  } else { # MODELobj$tinitx==0
    Vtt1.1 <- sub3D(Vtt1, t = 1)
    if(any(takediag(Vtt1.1)==0)){
      Vinv <- pcholinv(Vtt1.1, chol=FALSE)
      if (m != 1) Vinv <- symm(Vinv) # to enforce symmetry after chol2inv call
      J0 <- V00 %*% t.B %*% Vinv # eqn 6.49 and 1s on diag when Q=0; Here it is t.B[1]
    }else{
      t.J0 <- solve(matrix(Vtt1.1, m, m, byrow = TRUE), par.1$B%*%V00)
      if (m==1) J0 <- t.J0 else J0 <- matrix(t.J0, m, m, byrow = TRUE)
    }
    xtT.1 <- ks.out$alphahat[1:m, 1, drop = FALSE]
    x0T <- x00 + J0 %*% (ks.out$alphahat[1:m, 1, drop = FALSE] - ks.out$a[1:m, 1, drop = FALSE]) # eqn 6.47
    V0T <- V00 + tcrossprod(J0 %*% (VtT[, , 1] - Vtt1[, , 1]), J0) # eqn 6.48
    if (m != 1) V0T <- symm(V0T) # to enforce symmetry
  }
  
  if (!return.kfas.model) kfas.model <- NULL
  
  # not using ks.out$v (Innov) and ks.out$F (Sigma) since I think there might be a bug (or I misunderstand KFS output) when R is not diagonal.
  rtn.list <- list(
    xtT = ks.out$alphahat[1:m, , drop = FALSE],
    VtT = VtT,
    Vtt1T = Vtt1T,
    x0T = x0T,
    V0T = V0T,
    x10T = ks.out$alphahat[1:m, 1, drop = FALSE],
    V10T = V10T,
    x00T = x00,
    V00T = V00,
    Vtt = "Use MARSSkfss to get Vtt",
    Vtt1 = Vtt1,
    J = "Use MARSSkfss to get J",
    J0 = "Use MARSSkfss to get J0",
    Kt = "Use MARSSkfss to get Kt",
    xtt1 = ks.out$a[1:m, 1:TT, drop = FALSE],
    xtt = "Use MARSSkfss to get xtt",
    Innov = "Use MARSSkfss to get Innov",
    Sigma = "Use MARSSkfss to get Sigma",
    kfas.model = kfas.model,
    logLik = ks.out$logLik,
    ok = TRUE,
    errors = NULL
  )
  # apply names
  X.names <- attr(MODELobj, "X.names")
  for(el in c("xtT", "xtt1", "x0T", "x00T", "x10T")) rownames(rtn.list[[el]]) <- X.names
  return(rtn.list)
}
